tests/test_create_model.py

```diff
@@ -222,3 +222,27 @@ class TestGenericModel(GenericModel):
     result = AAModel[int](aa=1)
     assert result.aa == 1
     assert result.__config__.orm_mode is True
+
+
+def test_set_name():
+
+    from unittest.mock import Mock
+
+    from pydantic.fields import ModelPrivateAttr
+
+    mock = Mock()
+
+    class class_deco(ModelPrivateAttr):
+        def __init__(self, fn):
+            super().__init__()
+            self.fn = fn
+
+        def __set_name__(self, owner, name):
+            mock(owner, name)
+
+    class A(BaseModel):
+        @class_deco
+        def _some_func(self):
```

Comment:
can you also call `_some_func` to check it's a proper bound method.

Reply:
working on this one still, don't merge yet :)